symengine/dict.h

```diff
@@ -148,6 +152,119 @@ typedef struct {
 
 typedef std::unordered_map<vec_int, integer_class, vec_int_hash> umap_vec_mpz;
 
+int umap_vec_mpz_compare(const umap_vec_mpz &a, const umap_vec_mpz &b);
+long mpz_hash(const integer_class z);
+
+typedef std::vector<unsigned int> vec_uint;
+
+class vec_uint_hash
+{
+public:
+    std::size_t operator()(const vec_uint &v) const
+    {
+        std::size_t h = 0;
+        for (unsigned int i : v) {
+            h ^= i + 0x9e3779b + (h << 6) + (h >> 2);
+        }
+        return h;
+    }
+};
+
+class vec_uint_eq
+{
+public:
+    bool operator()(const vec_uint &a, const vec_uint &b) const
+    {
+        if (a.size() != b.size())
+            return false;
+        for (unsigned int i = 0; i < a.size(); i++) {
+            if (a[i] != b[i])
+                return false;
+        }
+        return true;
+    }
+};
+
+class vec_uint_compare
+{
+public:
+    bool operator()(const vec_uint &a, const vec_uint &b) const
+    {
+        if (a.size() != b.size())
+            return a.size() < b.size();
+        unsigned int sum1 = 0;
+        unsigned int sum2 = 0;
+        for (unsigned int x : a) {
+            sum1 += x;
+        }
+        for (unsigned int x : b) {
+            sum2 += x;
+        }
+        if (sum1 != sum2)
+            return sum1 < sum2;
+        return a < b;
+    }
+};
+
+typedef std::set<RCP<const Symbol>, RCPSymbolCompare> set_sym;
+typedef std::unordered_map<RCP<const Symbol>, unsigned int, RCPSymbolHash,
+                           RCPSymbolEq> umap_sym_uint;
+typedef std::unordered_map<vec_uint, integer_class, vec_uint_hash, vec_uint_eq>
+    umap_uvec_mpz;
+
+// Takes an unordered map of type M with key type K and returns a vector of K
+// ordered by C.
+template <class K, class M, class C>
+std::vector<K> order_umap(const M &d)
```

Comment:
This is returning an (ordered) vector of just the keys of the map? Is it supposed to be that way?
Also, using a `vector.insert` can take `O(N)` time, making this method take worst case `O(N^2)` time.

I suggest, you can create a `std::set` (or a `std::map`), which are both ordered. To this, pass your custom ordering `C` and key type `K`. Then iterate over all the entries in the unordered map, simply inserting the elements into the newly created set. Lastly, just flatten this set to a vector (there is a `std` function for this). This makes the method `O(NlogN)` in totality.